### ISSUE TITLE
1.6.0 Job initialization triggers a main thread disk read (Android JVM)

### ISSUE THREAD

**Issue by radixdev** [Kotlin/kotlinx.coroutines#3180]
[Android Strictmode](https://developer.android.com/reference/android/os/StrictMode) is stating that the initialization path of any created Job is triggering expensive disk reads (on the main thread). This behavior was introduced in 1.6.0 in [this PR](https://github.com/Kotlin/kotlinx.coroutines/pull/2957). 

Simply calling `.launch { }` on the CoroutineScope is enough to trigger this disk read issue. I propose that the eager uncaught exception logic is configurable behind a system property or in some way configurable to avoid this issue. Additionally, this code is running as part of an SDK so we cannot completely control when it is  executed as if we were the host application. 

@nreid260

# Custom CoroutineScope

```kotlin
object CompanyCoroutineScope : CoroutineScope {
    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
        companylog(E, e) { "Child job of CompanyCoroutineScope got exception: $e" }
    }

    override val coroutineContext: CoroutineContext = Dispatchers.IO + exceptionHandler + SupervisorJob()
}

```

# The disk reads
```
2022-02-09 15:05:07.077 21991-21991/com.company.sample D/StrictMode: StrictMode policy violation; ~duration=286 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1659)
        at java.io.UnixFileSystem.getLastModifiedTime(UnixFileSystem.java:289)
        at java.io.File.lastModified(File.java:937)
        at java.util.zip.ZipFile.<init>(ZipFile.java:265)
        at java.util.zip.ZipFile.<init>(ZipFile.java:187)
        at java.util.jar.JarFile.<init>(JarFile.java:169)
        at java.util.jar.JarFile.<init>(JarFile.java:106)
        at libcore.io.ClassPathURLStreamHandler.<init>(ClassPathURLStreamHandler.java:46)
        at dalvik.system.DexPathList$Element.maybeInit(DexPathList.java:750)
        at dalvik.system.DexPathList$Element.findResource(DexPathList.java:777)
        at dalvik.system.DexPathList.findResources(DexPathList.java:572)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:279)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:349)
        at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:402)
        at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:488)
        at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:786)
        at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:816)
        at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:807)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.<clinit>(CoroutineExceptionHandlerImpl.kt:23)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.initializeDefaultExceptionHandlers(CoroutineExceptionHandlerImpl.kt:27)
        at kotlinx.coroutines.Job.<clinit>(Job.kt:120)
        at kotlinx.coroutines.AbstractCoroutine.<init>(AbstractCoroutine.kt:51)
        at kotlinx.coroutines.StandaloneCoroutine.<init>(Builders.common.kt:190)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:55)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at com.company.images.DefaultCompanyImageLoader.initDiskCacheTask(DefaultCompanyImageLoader.kt:213)
        at com.company.images.DefaultCompanyImageLoader.<init>(DefaultCompanyImageLoader.kt:53)
        at com.company.Company.<init>(Company.java:230)
        at com.company.Company.<init>(Company.java:18)
        at com.company.Company.getInstance(Company.java:193)
        at com.company.Company.getInstance(Company.java:22)
        at com.company.sample.CompanyApplication.onCreate(CompanyApplication.kt:76)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1211)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6725)
        at android.app.ActivityThread.access$1500(ActivityThread.java:247)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2053)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
2022-02-09 15:05:07.078 21991-21991/com.company.sample D/StrictMode: StrictMode policy violation; ~duration=286 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1659)
        at java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:250)
        at java.io.File.isDirectory(File.java:843)
        at java.io.File.toURI(File.java:728)
        at libcore.io.ClassPathURLStreamHandler.<init>(ClassPathURLStreamHandler.java:51)
        at dalvik.system.DexPathList$Element.maybeInit(DexPathList.java:750)
        at dalvik.system.DexPathList$Element.findResource(DexPathList.java:777)
        at dalvik.system.DexPathList.findResources(DexPathList.java:572)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:279)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:349)
        at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:402)
        at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:488)
        at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:786)
        at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:816)
        at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:807)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.<clinit>(CoroutineExceptionHandlerImpl.kt:23)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.initializeDefaultExceptionHandlers(CoroutineExceptionHandlerImpl.kt:27)
        at kotlinx.coroutines.Job.<clinit>(Job.kt:120)
        at kotlinx.coroutines.AbstractCoroutine.<init>(AbstractCoroutine.kt:51)
        at kotlinx.coroutines.StandaloneCoroutine.<init>(Builders.common.kt:190)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:55)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at com.company.images.DefaultCompanyImageLoader.initDiskCacheTask(DefaultCompanyImageLoader.kt:213)
        at com.company.images.DefaultCompanyImageLoader.<init>(DefaultCompanyImageLoader.kt:53)
        at com.company.Company.<init>(Company.java:230)
        at com.company.Company.<init>(Company.java:18)
        at com.company.Company.getInstance(Company.java:193)
        at com.company.Company.getInstance(Company.java:22)
        at com.company.sample.CompanyApplication.onCreate(CompanyApplication.kt:76)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1211)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6725)
        at android.app.ActivityThread.access$1500(ActivityThread.java:247)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2053)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
2022-02-09 15:05:07.078 21991-21991/com.company.sample D/StrictMode: StrictMode policy violation; ~duration=285 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1659)
        at java.io.UnixFileSystem.getLastModifiedTime(UnixFileSystem.java:289)
        at java.io.File.lastModified(File.java:937)
        at java.util.zip.ZipFile.<init>(ZipFile.java:265)
        at java.util.zip.ZipFile.<init>(ZipFile.java:187)
        at java.util.jar.JarFile.<init>(JarFile.java:169)
        at java.util.jar.JarFile.<init>(JarFile.java:106)
        at libcore.io.ClassPathURLStreamHandler.<init>(ClassPathURLStreamHandler.java:46)
        at dalvik.system.DexPathList$Element.maybeInit(DexPathList.java:750)
        at dalvik.system.DexPathList$Element.findResource(DexPathList.java:777)
        at dalvik.system.DexPathList.findResources(DexPathList.java:572)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:279)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:290)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:349)
        at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:402)
        at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:488)
        at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:786)
        at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:816)
        at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:807)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.<clinit>(CoroutineExceptionHandlerImpl.kt:23)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.initializeDefaultExceptionHandlers(CoroutineExceptionHandlerImpl.kt:27)
        at kotlinx.coroutines.Job.<clinit>(Job.kt:120)
        at kotlinx.coroutines.AbstractCoroutine.<init>(AbstractCoroutine.kt:51)
        at kotlinx.coroutines.StandaloneCoroutine.<init>(Builders.common.kt:190)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:55)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at com.company.images.DefaultCompanyImageLoader.initDiskCacheTask(DefaultCompanyImageLoader.kt:213)
        at com.company.images.DefaultCompanyImageLoader.<init>(DefaultCompanyImageLoader.kt:53)
        at com.company.Company.<init>(Company.java:230)
        at com.company.Company.<init>(Company.java:18)
        at com.company.Company.getInstance(Company.java:193)
        at com.company.Company.getInstance(Company.java:22)
        at com.company.sample.CompanyApplication.onCreate(CompanyApplication.kt:76)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1211)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6725)
        at android.app.ActivityThread.access$1500(ActivityThread.java:247)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2053)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
2022-02-09 15:05:07.079 21991-21991/com.company.sample D/StrictMode: StrictMode policy violation; ~duration=285 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1659)
        at java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:250)
        at java.io.File.isDirectory(File.java:843)
        at java.io.File.toURI(File.java:728)
        at libcore.io.ClassPathURLStreamHandler.<init>(ClassPathURLStreamHandler.java:51)
        at dalvik.system.DexPathList$Element.maybeInit(DexPathList.java:750)
        at dalvik.system.DexPathList$Element.findResource(DexPathList.java:777)
        at dalvik.system.DexPathList.findResources(DexPathList.java:572)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:279)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:290)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:349)
        at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:402)
        at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:488)
        at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:786)
        at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:816)
        at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:807)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.<clinit>(CoroutineExceptionHandlerImpl.kt:23)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.initializeDefaultExceptionHandlers(CoroutineExceptionHandlerImpl.kt:27)
        at kotlinx.coroutines.Job.<clinit>(Job.kt:120)
        at kotlinx.coroutines.AbstractCoroutine.<init>(AbstractCoroutine.kt:51)
        at kotlinx.coroutines.StandaloneCoroutine.<init>(Builders.common.kt:190)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:55)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at com.company.images.DefaultCompanyImageLoader.initDiskCacheTask(DefaultCompanyImageLoader.kt:213)
        at com.company.images.DefaultCompanyImageLoader.<init>(DefaultCompanyImageLoader.kt:53)
        at com.company.Company.<init>(Company.java:230)
        at com.company.Company.<init>(Company.java:18)
        at com.company.Company.getInstance(Company.java:193)
        at com.company.Company.getInstance(Company.java:22)
        at com.company.sample.CompanyApplication.onCreate(CompanyApplication.kt:76)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1211)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6725)
        at android.app.ActivityThread.access$1500(ActivityThread.java:247)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2053)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

**Comment by qwwdfsad** [comment]
Could you please elaborate whether you have R8 enabled?

We previously had the same issue with `Dispatchers.Main` (#1231), but it was fixed with specific R8 optimization

**Comment by qwwdfsad** [comment]
If it is okay for you, could you please file small reproducing project so we can dissect it?

**Comment by radixdev** [comment]
Hi @qwwdfsad ,

Apologies for the weekend delay. https://github.com/radixdev/sscs-kotlin-coroutines-3180 is a reproducing project.

**Comment by qwwdfsad** [comment]
Thanks! Verified my fix works

**Comment by qwwdfsad** [comment]
Fixed in 1.6.1

**Comment by radixdev** [comment]
Thank you @qwwdfsad !
